### ISSUE TITLE
Improved error message for unmatched inputs

### ISSUE THREAD

**Issue by spencergilbert** [vectordotdev/vector#14635]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Use Cases

Today the error message a user gets when the `input` of a downstream component isn't matched is:

```text
ERROR vector::cli: Configuration error. error=Input "datadog_agents" for transform "remap_logs_for_datadog" doesn't match any components.
```

This is because `datadog_agents` is using multiple outputs, so the names for `input` are actually: `datadog_agents.logs`, `datadog_agents.metrics`, `datadog_agents.traces`. This will only continue to be a confusing experience as more multiple output components are added.

### Attempted Solutions

There's no existing workaround, aside from user education and improved documentation, today.

### Proposal

We could compare the unmatched input to our known list of names and determine if any are "close"  (for some definition of close). We could then provide more useful feedback to the operator.

```text
ERROR vector::cli: Configuration error. error=Input "datadog_agents" for transform "remap_logs_for_datadog" doesn't match any components. Did you mean "datadog_agents.logs"?
```

Related example:

```shell
❯ cargo clear
error: no such subcommand: `clear`

	Did you mean `clean`?
```

### Version

vector 0.24.1 (x86_64-apple-darwin 8935681 2022-09-12)

**Comment by hhromic** [comment]
This is a _really great_ idea!
Regarding the measure of string "closeness" to use, I think a good starting point could be the [Levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance), which mostly measures how many _edits_ a string needs (the user input) to be equal to another string (all known output names).

Another option is to just go check what string comparison metric `cargo` uses and be Rust-consistent :)

Nevertheless, this would be a really nice improvement to the UX with Vector.
Smarter error reporting = happier developers! (Looking at you C++ errors 👀)

**Comment by spencergilbert** [comment]
> Nevertheless, this would be a really nice improvement to the UX with Vector. Smarter error reporting = happier developers! (Looking at you C++ errors 👀)
> 
> EDIT: An idea like this could be applied to more than just output names, for example config keys :)

Yeah, it'd be great if we could provide a generic solution that would work _wherever_ in Vector. We've gotten a lot of compliments around the UX/feedback in VRL (in terms of "rustc-like" compiler errors), it'd be nice if we could extend that into the rest of Vector proper.
